README.md
# Purpose

The purpose of the Soil Organic Carbon Data Rescue and Harmonization Repository (SOC-DRaHR) is to collect haromonization scripts for data sets related to soil carbon to enable scientific research.
This project identifies soil carbon datasets that are publicly avaiable, provides data harmonization scripts to integrate those data sets, and provides output scripts for a harmonized data product.
Data can come from field surveys, field manipulation studies, and laboratory experiments.

This project is not a _data_ repository or archive but instead a code repository.
End users are responsible for complying with ALL data use policies of the orginal data providers, please check with the orginal archives and reposities to ensure you are complying with use policies.

# Project links

Here are some important resources to get you started:
1) [International Soil Carbon Network](http://iscn.fluxdata.org/) is our parent organization.
2) [Our code of conduct](CONTRIBUTING.md) is based on the [The Contributor Covenant](https://www.contributor-covenant.org/).
3) [Our roadmap](roadmap.md) of where we want the project to go currently and over the next 1, 5, and 10 years.
4) [How to contribute](CONTRIBUTING.md)
5) [Our mailing list](https://groups.google.com/forum/#!forum/soc-drahr)

# Quick start

To quickly access ISCN3 data:
```R
library(devtools)
install_github("ISCN/soilDataR")
library(SoilDataR)
ISCN3 <- processData_ISCN3()
```

See [ISCNVisulaize](https://github.com/ktoddbrown/SOC-DRaHR/blob/master/reports/ISCNVisualize.Rmd) for a detailed breakdown of the variables and site locations in ISCN3.

# How to contribute
Please see the [CONTRIBUTING](CONTRIBUTING.md) document for more details on how to contribute, including how to *identify* datasets, contribute to the *code*, and *everything else* that is needed to run an open source community project.
Contributers will be offered co-authorship for repository DOI but can not be guaranteed co-authorship on manuscripts, proposals or studies that utilize this repository.
[This repository is licensed here](LICENSE).

# Open data calls
Below are active data calls for future manuscripts.

1) Labratory incubated soil respiration response to moisture manipulation. Reviewers: K Todd-Brown and B Bond-Lamberty

# Citations
Repository versions will be assigned DOIs as needed.
Any manuscripts which use these data harmonization scripts are asked to cite the appropreate repository version but are not required to list contributers as co-authors.
The aggregation scripts here are licensed under BSD 2-clause. See [LICENSE](LICENSE.txt) for details.

# Outside links of interest

[Environmental Data Initiative](https://environmentaldatainitiative.org/) has a [GitRepo](https://github.com/EDIorg/ecocomDP) that has a similar approach to aggregating survey data.

[Earth Sciences Information Partners](http://www.esipfed.org/)

[DataOne](https://www.dataone.org/) is a go-to resource for finding environmental data in general and searches many different established repositories. They have an R package for interacting with their search engine [here](https://jsta.github.io/2017/03/28/dataone.html)

[rOpenSci Project](https://ropensci.github.io/)

Five ways consortia can catalyse open science, [Nature, 2017](http://www.nature.com/news/five-ways-consortia-can-catalyse-open-science-1.21706)

[National Data Service](http://www.nationaldataservice.org/)


news.md
|Date | |
|-----|---|
|February 6, 2018 | Repository transfered from personal to ISCN organization repo |
|December 10, 2017 | Our first hackaton in New Orleans, LA! Held the Sunday before AGU. |
|November 14, 2017 | Harmonize data sets (Treat and Alamos) that were queued up for ISCN 3.1 data ingest. |
|May 25, 2017 | Comments on variables defined in ISCN 3 added to the ISCN report. |
|Febuary 27, 2017 | This git repository is live! |


repoData/ISCN_3/processData_ISCN3.R
library(xlsx)
library(reshape2)
library(plyr)
processData_ISCN3 <- function(dir='repoData/ISCN_3'){
  
  source(sprintf('%s/processMethodBlock.R', dir))
  source(sprintf('%s/processWorksheet.R', dir))
  
  data1.ls <- processWorksheet(csvFile=
                                 sprintf('%s/Layers/ISCN_ALL_DATA_LAYER_C1_1-1.csv', dir),
                               verbose=TRUE)
  
  data2.ls <- processWorksheet(csvFile=
                                 sprintf('%s/Layers/ISCN_ALL_DATA_LAYER_C2_1-1.csv', dir),
                               verbose=TRUE)
  
  data3.ls <- processWorksheet(csvFile=
                                 sprintf('%s/Layers/ISCN_ALL_DATA_LAYER_C3_1-1.csv', dir),
                               verbose=TRUE)
  
  data4.ls <- processWorksheet(csvFile=
                                 sprintf('%s/Layers/ISCN_ALL_DATA_LAYER_C4_1-1.csv', dir),
                               verbose=TRUE)
  #merge easy========
  study.df <- data1.ls$study
  fieldTreatment.df <- data1.ls$fieldTreatment
  labTreatment.df <- data1.ls$labTreatment
  
  lab.df <- unique(rbind.fill(rbind.fill(rbind.fill(data1.ls$lab,data2.ls$lab), 
                                         data3.ls$lab), data4.ls$lab))
  
  field.df <- unique(rbind.fill(rbind.fill(rbind.fill(data1.ls$field,data2.ls$field), 
                                           data3.ls$field), data4.ls$field))
  field.df$value <- as.factor(field.df$value)
  field.df$layer_units <- as.factor(field.df$layer_units)
  
  
  fieldNotes.df <- field.df[,c('fieldID', 'measurement', 'value')]
  field.df <- unique(field.df[,c('fieldID', 'lat', 'long', 'observation_date', 
                                 'layer_top', 'layer_bot', 'layer_units')])
  #mergeMeasures====
  measureTemp.df <- 
    merge(merge(data1.ls$measurement, data2.ls$measurement, by=c('type', 'method'), 
                suffixes=c('.data1', '.data2'), all=TRUE),
          merge(data3.ls$measurement, data4.ls$measurement, by=c('type', 'method'), 
                suffixes=c('.data3', '.data4'), all=TRUE), all=TRUE)
  
  measureTemp.df <- ddply(measureTemp.df, c('type'), function(xx){
    xx$measurementID <- sprintf('%s_%02d', xx$type, 1:nrow(xx))
    return(xx)
  })
  measurement.df <- measureTemp.df[,c('type', 'method', 'measurementID')]
  
  #Construct samples========
  sample.df <- ddply(data1.ls$samples, 'measurementID', function(xx){
    #cat('replacing [', xx$measurementID[1])
    xx$measurementID <- subset(measureTemp.df,
                               xx$measurementID[1] == measurementID.data1)$measurementID
    #cat('] with [',xx$measurementID[1],']\n')
    return(xx)
  })
  
  sample.df <- rbind.fill(sample.df, 
                          ddply(data2.ls$samples, 'measurementID', function(xx){
                            #cat('replacing [', xx$measurementID[1])
                            xx$measurementID <- subset(measureTemp.df,
                                                       xx$measurementID[1] ==
                                                         measurementID.data2)$measurementID
                            #cat('] with [',xx$measurementID[1],']\n')
                            return(xx)
                          }))
  
  sample.df <- rbind.fill(sample.df, 
                          ddply(data3.ls$samples, 'measurementID', function(xx){
                            #cat('replacing [', xx$measurementID[1])
                            xx$measurementID <- subset(measureTemp.df,
                                                       xx$measurementID[1] == 
                                                         measurementID.data3)$measurementID
                            #cat('] with [',xx$measurementID[1],']\n')
                            return(xx)
                          }))
  
  sample.df <- rbind.fill(sample.df, 
                          ddply(data4.ls$samples, 'measurementID', function(xx){
                            #cat('replacing [', xx$measurementID[1])
                            xx$measurementID <- subset(measureTemp.df,
                                                       xx$measurementID[1] == measurementID.data4)$measurementID
                            #cat('] with [',xx$measurementID[1],']\n')
                            return(xx)
                          }))
  sample.df$fieldID <- as.factor(sample.df$fieldID)
  sample.df$measurementID <- as.factor(sample.df$measurementID)
  sample.df$unit <- as.factor(sample.df$unit)
  
  return(list(study=study.df, labTreatment=labTreatment.df, fieldTreatment=fieldTreatment.df,
              field=field.df, fieldNotes=fieldNotes.df, measurement=measurement.df, 
              sample=sample.df))
}

repoData/ISCN_3/processMethodBlock.R
#' Process a methods block
#' 
#' Force the samples to discard repeats
#'
#' @param methodNames 
#' @param unitName 
#' @param sampleTemp 
#' @param unit.df 
#'
#' @return
#' @export
#'
#' @examples
processMethodBlock <- function(methodNames=NULL, unitName=NULL,
                               sampleTemp, unit.df=NULL, verbose=FALSE){
  
  ##Merge methods
  sampleTemp$method <- ''
  for(methodStr in methodNames){
    patternStr <- sprintf('%%s %s: %%s', methodStr)
    emptyStr <- sprintf('\\s+%s:\\s+$', methodStr)
    sampleTemp$method <- sprintf(patternStr, sampleTemp$method, sampleTemp[,methodStr])
    sampleTemp$method <- gsub(emptyStr, '', sampleTemp$method)
    sampleTemp$method <- gsub('^\\s+', '', sampleTemp$method)
    sampleTemp[,methodStr] <- NULL
  }
  if(verbose) cat('flag1')
  #melt values
  sampleTemp <- unique(melt(sampleTemp, id.vars=c(c('fieldID', 'method'), unitName), 
                     variable.name='measurement', 
                     na.rm=TRUE, forceNumericValue=TRUE))
  if(verbose) cat('flag2')
  if(!is.null(unitName)){
    names(sampleTemp)[3] <- 'unit'
  }else if(is.null(unitName) & !is.null(unit.df)){
    #units
    sampleTemp <- merge(sampleTemp, unit.df[,c('measurement', 'unit')])
  }else{
    warning('units not specified')
  }
  if(verbose) cat('flag3')
  #issolate unique measurements
  measurementTemp <- unique(sampleTemp[c('measurement','method')])
  measurementTemp$measurementID <- sprintf('%s_%02d', measurementTemp$measurement, 
                                           1:nrow(measurementTemp))
  if(verbose) cat('flag4')
  sampleTemp <- merge(sampleTemp, measurementTemp)[, c('fieldID', 'measurementID',
                                                       'value', 'unit')]
  
  #rename stuff
  names(measurementTemp)[1] <- 'type'
  
  sampleTemp$value <- as.numeric(sampleTemp$value)
  if(verbose) cat('flag5')
  sampleTemp <- sampleTemp[sampleTemp$value != -999, ]
  if(verbose) cat('flag6')
  return(list(sample=sampleTemp, measurement=measurementTemp))
  
}

repoData/Hardon2005/processData_Hardon2005.R
library(XML)
library(reshape2)
library(plyr)
processData_Hardon2005 <- function(dir='repoData/'){
  xml_tree <- xmlTreeParse(sprintf('%sHardon2005/MetaData.xml', dir)) 
  xlm_ls <- xmlToList(sprintf('%sHardon2005/MetaData.xml', dir))
  #columns: xlm_ls$dataset$dataTable$attributeList[[1]]
  
  ##Pull site information
  fieldData <- data.frame()
  for(site.ls in xlm_ls$dataset$methods$sampling$studyExtent$coverage){
    if('boundingCoordinates' %in% names(site.ls)){
      fieldData <- rbind(fieldData,
                    data.frame(fieldTreatmentID = unlist(strsplit(site.ls$geographicDescription, 
                                                      split=' -- '))[1],
                                    description=unlist(strsplit(site.ls$geographicDescription,
                                                                split=' -- '))[2],
                                    lat=site.ls$boundingCoordinates$northBoundingCoordinate,
                                    lon=site.ls$boundingCoordinates$westBoundingCoordinate))
    }
  }
  fieldData$fieldName <- 'Bonanza Creek LTER'
  
  fieldTreatment <- fieldData[, c('fieldTreatmentID', 'description')]
  fieldData <- unique(fieldData[, c('fieldName', 'lat', 'lon')])
  
  ##Pull creator information
  labData <- data.frame()
  for(personel.ls in xlm_ls$dataset[names(xlm_ls$dataset) %in% 'creator']){
    personelInfo <- c("individualName.givenName",
                      "individualName.surName",   "positionName",
                      "electronicMailAddress",    "onlineUrl")
    orgInfo <- c('organizationName', 'electronicMailAddress', 'onlineUrl')
    if(all(personelInfo %in% names(unlist(personel.ls)))){
      labData <- rbind.fill(labData,
                            data.frame(t(unlist(personel.ls)[personelInfo])))
    }else if(all(orgInfo %in% names(unlist(personel.ls)))){
      labData <- rbind.fill(labData,
                            data.frame(t(unlist(personel.ls)[orgInfo])))
    }
  }
  labData$lab_ID <- 'Harden-Turetsky'
  
  ##Pull column names and information
  tableMeta <- data.frame()
  for(column.ls in xlm_ls$dataset$dataTable$attributeList){
    tableMeta <- rbind(tableMeta,
                       data.frame(Name=column.ls$attributeName, Label=column.ls$attributeLabel,
                                  Definition=column.ls$attributeDefinition, 
                                  Type=column.ls$storageType,
                                  Scale=paste0(unlist(column.ls$measurementScale), 
                                               collapse=' - '),
                                  comment=ifelse(is.null(column.ls$comment), '', 
                                                 column.ls$comment)))
  }
  tableMeta$columnIndex <- paste('column', row.names(tableMeta), sep='_')
  
  
  data.df <- read.csv(sprintf('%sHardon2005/335_BNZ_moisturegradient_isotopestudy_2005_LTER.txt',
                              dir))
  
  #human check the naming
  print(rbind(as.character(tableMeta$Name), names(data.df)))
  
  names(data.df)[ 7:ncol(data.df)] <- paste('column', 7:ncol(data.df), sep='_')
  
  data.df <- melt(data.df, id.vars=names(data.df)[1:6], variable.name='columnIndex')
  
  tableMeta$IncubationTime <- NA
  tableMeta$IncubationTime[grepl('\\d+ ((hours)|(hrs)|(days))', as.character(tableMeta$Name))] <-
    regmatches(as.character(tableMeta$Name), regexpr('\\d+ ((hours)|(hrs)|(days))', 
                                                     as.character(tableMeta$Name)))
  tableMeta$IncubationTime[grepl('initial', tableMeta$Name)] <- '2 days'
  tableMeta$IncubationTime.units <- NA
  tableMeta$IncubationTime.units[grepl('\\d+ ((hours)|(hrs)|(days))', 
                                       as.character(tableMeta$IncubationTime))] <- 
    regmatches(as.character(tableMeta$IncubationTime), regexpr('((hours)|(hrs)|(days))', 
                                                    as.character(tableMeta$IncubationTime)))
  tableMeta$IncubationTime <- gsub(' ((hours)|(hrs)|(days))', '', 
                                   as.character(tableMeta$IncubationTime))
  tableMeta$measure_id <- paste(ifelse(grepl('14C',tableMeta$Name), 'd14C_', ''), 
                                paste(ifelse(grepl('13C',tableMeta$Name), 'd13C_', ''), 
                                      ifelse(grepl('(CO2)|(14C)', tableMeta$Name),
                                             'CO2_flux_lab_48hr', ''), sep=''), sep='')
  tableMeta$measure_units <- ifelse(grepl('^d',tableMeta$measure_id), 'ppm', 
                               ifelse(grepl('ppm/d', tableMeta$Scale), 'ppm cm-soil^-3 day^-1',
                                           ifelse(grepl('mgCO2/d', tableMeta$Scale), 
                                                  'mg-CO2 cm-soil^-3 day^-1',
                                                  '')))
  
  soilVolume <-  5*pi*(4.25/2)^2 #(From methods) preserved core size 5 cm of soil using a 4.25 cm
  
  data.df <- merge(data.df, tableMeta, by='columnIndex')
  data.df$value <- data.df$value/(grepl('cm-soil', data.df$measure_units)*soilVolume + 
                                    !grepl('cm-soil', data.df$measure_units))
  data.df$DOI <- '10.6073/pasta/9b363cc74bcd784cfcee85e6920f2989'
  data.df$Warming.status <- ifelse(grepl('warm', data.df$Warming.status), 
                                   'field_warming_NAC', 'control')
  
  #time (collection date), depth.min, depth.max, depth.unit
  fieldData$'mm-dd-yyyy' <- unique(data.df$Sampling.Date)
  fieldData$depth.unit <- 'cm'
  #fieldData <- rbind(fieldData, fieldData) #quick and dirty hack
  fieldData <- data.frame(fieldData,
                          Depth=as.character(unique(data.df$Depth)),
                          depth.min = as.numeric(as.character(unlist(data.frame(strsplit(
                            as.character(unique(data.df$Depth)), split='-'))[1,]))),
                          depth.max = as.numeric(as.character(unlist(data.frame(strsplit(
                            as.character(unique(data.df$Depth)), split='-'))[2,]))))
  fieldData$fieldID <- paste('BC-LTER', fieldData$Depth, sep='_')
  
  fieldTreatment <- merge(fieldTreatment, unique(data.df[, c('Warming.status', 'Site')]), 
                          by.x='fieldTreatmentID', by.y='Site', all=TRUE)
  fieldTreatment$vegID <- fieldTreatment$fieldTreatmentID
  fieldTreatment$Site <- fieldTreatment$vegID
  fieldTreatment$fieldTreatmentID <- with(fieldTreatment, 
                                          sprintf('%s_%s', fieldTreatmentID, Warming.status))
  
  data.df <- merge(data.df, fieldTreatment)
  data.df$lab_ID <- unique(labData$lab_ID) #Only one ID
  data.df <- merge(data.df, fieldData)
  
  return(list(data=data.df[c('lab_ID', 'fieldID', 'fieldTreatmentID', 'Sample.ID', 
                             'IncubationTime', 
                             'IncubationTime.units', 'measure_id', 'value', 'measure_units')],
              field=fieldData,
              fieldTreatment=fieldTreatment,
              lab=labData,
              labTreatment=NA))
}

hackathon/2017_December_NewOrleans/processData_BES-LTER.R
## procesData_BES-LTER.R
#' Script to process soil dataset from BES-LTER site
#' Created by Sam Zipper during AGU 2017 Soil Hackathon
#' 
#' Data description: https://beslter.org/metacat_harvest_attribute_level_eml/html_metadata/bes_584.asp
#' Data: http://beslter.org/pub/soil/chemistry/RC_Summary_Data-AddressesReplacedWithNumbers.txt

require(ggplot2)
require(dplyr)

df.in <- read.csv("http://beslter.org/pub/soil/chemistry/RC_Summary_Data-AddressesReplacedWithNumbers.txt", stringsAsFactors=F)

# make site dataframe -----------------------------------------------------

df.site.temp <- unique(df.in[,c("Site", "LU_Current", "LU_Previous", "Yr_Built")])

# note: previous vegtype and year of conversion
# only available for non-forest sites
df.site <- data.frame(site_name = df.site.temp$Site,
                      vegtype = df.site.temp$LU_Current,
                      veg_note = paste0("Previous veg type: ", df.site.temp$LU_Previous, "; converted ", df.site.temp$Yr_Built))

# set lat/lon - note that site coordinates are not provided to preserve anonymity of landowners
# but all of them are within a 17 km^2 watershed with bounding box: 
# West	-77.314183; East	-76.012008; North	39.724847; South	38.708367
df.site$lat <- -76.6631
df.site$lon <- 39.21661
df.site$location_acc <- 17*(1000^2)

rm(df.site.temp)

# make metadata dataframe --------------------------------------------------

df.metadata <- data.frame(curator_personName="Jonathan Walsh",
                          curator_organization="Baltimore Ecosystem Study LTER",
                          curator_email="walshj@caryinstitute.org",
                          dataset_name="Physical, chemical and biological properties of forest and home lawn soils",
                          dataset_description = "One-meter soil cores were taken to evaluate soil texture, bulk density, carbon and nitrogen pools, microbial biomass carbon and nitrogen content, microbial respiration, potential net nitrogen mineralization, potential net nitrification and inorganic nitrogen pools in 32 residential home lawns that differed by previous land use and age, but had similar soil types. These were compared to soils from 8 forested reference sites.",
                          reference = "Raciti, S. R., P. M. Groffman, J. C. Jenkins, R. V. Pouyat, and T. J. Fahey. 2011a. Controls on nitrate production and availability in residential soils. Ecological Applications:In press. Raciti, S. R., P. M. Groffman, J. C. Jenkins, R. V. Pouyat, T. J. Fahey, M. L. Cadenasso, and S. T. A. Pickett. 2011b. Accumulation of carbon and nitrogen in residential soils with different land use histories. Ecosystems 14:287-297.",
                          acknowledgment = "The scientists the BES-LTER Program and the Cary Institute of Ecosystem Studies must receive adequate acknowledgement for the use of data. Please inform the BES Project Facilitator and use a statement such as the following to give credit on any document using the data: Data was supported by the Baltimore Ecosystem Study")

# make sample dataframe ---------------------------------------------------

df.sample <- data.frame(site_name = df.in$Site,
                        sample = df.in$REP.,
                        layer_top = 0,
                        layer_bottom = 10)

df.sample$layer_top[df.in$Depth=="0to10"] <- 0
df.sample$layer_bottom[df.in$Depth=="0to10"] <- 10

df.sample$layer_top[df.in$Depth=="10to30"] <- 10
df.sample$layer_bottom[df.in$Depth=="10to30"] <- 30

df.sample$layer_top[df.in$Depth=="30to70"] <- 30
df.sample$layer_bottom[df.in$Depth=="30to70"] <- 70

df.sample$layer_top[df.in$Depth=="70to100"] <- 70
df.sample$layer_bottom[df.in$Depth=="70to100"] <- 100

# bulk density
df.sample$bd_samp <- df.in$BD
df.sample$bd_method <- "Subsamples of homogenized soil from each depth interval were analyzed for soil dry weight and percent moisture (48 hrs at 105 oC). Bulk density (BD) was calculated as BD = (Total Dry Mass - Rock Mass) / (Total Volume - Rock Volume)."

# soil texture
df.sample$sand_tot_psa <- df.in$Sand_Perc
df.sample$silt_tot_psa <- df.in$Silt_Perc
df.sample$clay_tot_psa <- df.in$Clay_Perc

# carbon and nitrogen
df.sample$c_tot <- df.in$C_Perc
df.sample$n_tot <- df.in$N_Perc
df.sample$c_to_n <- df.in$C_N

# transform sample from wide to long format
df.sample.long <- reshape2::melt(df.sample, id=c("site_name", "sample", "layer_top", "layer_bottom", "bd_method"))


repoData/ISCN_3/processWorksheet.R
processWorksheet <- function(csvFile='Layers/ISCN_ALL_DATA_LAYER_C1_1-1.csv',
                             verbose=TRUE){
  ##Out of memory erros
  #header <- read.xlsx('Layers/ISCN_ALL_DATA_LAYER_C1_1-1.xlsx', sheetIndex=1, startRow=1, endRow=2)
  #data.df <- read.xlsx2('Layers/ISCN_ALL_DATA_LAYER_C1_1-1.xlsx', sheetIndex=1)
  
  #Load file========
  if(verbose) cat('loading csv:', csvFile, '\n')
  data.df <- read.csv(csvFile, stringsAsFactors=FALSE)
  if(verbose) cat('inital size: ', format(object.size(data.df), units='Mb'), '\n')
  
  #trim replicates
  #data.df <- subset(data.df, !grepl('ISCN SOC stock', dataset_name_soc))
  #if(verbose) cat('trim ISCN SOC stocks: ', format(object.size(data.df), units='Mb'), '\n')
  
  data.df <- unique(data.df)
  if(verbose) cat('trim non-unique: ', format(object.size(data.df), units='Mb'), '\n')
  
  #Process header column =============
  ##process the header for units
  if(verbose) cat('setting up header\n')
  header <- read.csv(csvFile, nrows=1, header=FALSE)
  
  header.df <- data.frame(header=unlist(header),
                          measurement=gsub(' \\(.*\\)', '', unlist(header)),
                          unit=gsub('\\)', '', gsub('.* \\(', '', unlist(header))),
                          stringsAsFactors=FALSE)
  header.df$unit[header.df$measurement == header.df$unit] <- NA
  header.df$unit[grepl('ph_(cacl|h2o|other)', header.df$measurement)] <- 'unitless'
  header.df$unit[grepl('root_quant_size', header.df$measurement)] <- 'unitless'
  header.df$unit[92:95] <- 'unitless'
  
  #Set up study, lab, and treatment ===========
  ##Set up the easy data frame
  if(verbose) cat('reading study/lab/treatment information\n')
  study.df <- data.frame(studyID = header[1], doi='10.17040/ISCN/1305039', permissions='acknowledgement')
  
  lab.df <- data.frame(labID=unique(data.df[,2])) #ignore dataset_name_SOC, back out SOC later
  
  fieldTreatment.df <- data.frame(fieldTreatmentID=NA) #no field treatment
  labTreatment.df <- data.frame(labTreatmentID=NA) #no lab treatment
  
  #set up field IDs============
  if(verbose) cat('setting up field ID\n')
  field.df <- unique(data.df[,c(4:22, c(84, 94, 95))])
  names(field.df) <- as.character(header.df$measurement[c(4:22, c(84, 94, 95))])
  field.df$fieldID <- field.df$layer_name
  field.df <- melt(field.df, id.vars=c('fieldID', 'lat', 'long', 'observation_date', 
                                       'layer_top', 'layer_bot'), 
                   variable.name='measurement', factorsAsStrings=TRUE)
  field.df$layer_units <- 'cm'
  field.df$layer_top <- as.numeric(as.character(field.df$layer_top))
  field.df <- field.df[!grepl('^\\s*$',field.df$value),]
  
  ####Pull data subsets, sampleID = fieldID = layer_name [12]
  if(verbose) cat('pulling data\n')
  #BulkDensity===================
  if(verbose) cat('Bulk density \n')
  ##Bulk density columns 23:28
  #header.df[c(12, 23:28),]
  sampleTemp <- data.df[, c(12, 23:28)]
  names(sampleTemp) <- header.df$measurement[c(12,23:28)]
  names(sampleTemp)[1] <- 'fieldID'
  
  temp <- processMethodBlock(methodNames=header.df$measurement[c(23, 28)],
                             sampleTemp=sampleTemp, 
                             unit.df=header.df[c(24:27), c('measurement', 'unit')],
                             verbose=FALSE)
  
  #merge results
  measurement.df <- temp$measurement
  sample.df <- temp$sample
  
  #CARBON=====================
  if(verbose) cat('Carbon \n')
  ##Carbon, columns 29-33
  #header.df[c(12, 29:33),]
  sampleTemp <- data.df[, c(12, 29:33)]
  names(sampleTemp) <- header.df$measurement[c(12,29:33)]
  names(sampleTemp)[1] <- 'fieldID'
  headerTemp <- 
  
  temp <- processMethodBlock(methodNames=header.df$measurement[29:30],
                             sampleTemp=sampleTemp, 
                             unit.df=header.df[c(31:33), c('measurement', 'unit')])
  #merge results
  measurement.df <- rbind.fill(measurement.df, temp$measurement)
  sample.df <- rbind.fill(sample.df, temp$sample)
  
  #Nitrogen=============================
  if(verbose) cat('Nitrogen \n')
  ##Nitrogen, columns 34-35
  #header.df[c(12, 34:35),]
  sampleTemp <- data.df[, c(12, 34:35)]
  names(sampleTemp) <- header.df$measurement[c(12,34:35)]
  names(sampleTemp)[1] <- 'fieldID'
  temp <- processMethodBlock(methodNames=NULL,
                             sampleTemp=sampleTemp, 
                             unit.df=header.df[c(34:35), c('measurement', 'unit')])
  #merge results
  measurement.df <- rbind.fill(measurement.df, temp$measurement)
  sample.df <- rbind.fill(sample.df, temp$sample)
  
  #SOC==================
  if(verbose) cat('SOC\n')
  ##SOC, columns 36-38
  #header.df[c(12, 36:38),]
  sampleTemp <- data.df[, c(12, 36:38)]
  names(sampleTemp) <- header.df$measurement[c(12,36:38)]
  names(sampleTemp)[1] <- 'fieldID'
  temp <- processMethodBlock(methodNames=header.df$measurement[c(37:38)],
                             sampleTemp=sampleTemp, 
                             unit.df=header.df[c(36), c('measurement', 'unit')])
  #merge results
  measurement.df <- rbind.fill(measurement.df, temp$measurement)
  sample.df <- rbind.fill(sample.df, temp$sample)
  
  ##pH, columns 39-42
  if(verbose) cat('pH\n')
  #header.df[c(12, 39:42),]
  sampleTemp <- data.df[, c(12, 39:42)]
  names(sampleTemp) <- c('fieldID', header.df$measurement[c(39:42)])
  temp <- processMethodBlock(methodNames=header.df$measurement[c(39)],
                             sampleTemp=sampleTemp, 
                             unit.df=header.df[c(40:42), c('measurement', 'unit')])
  #merge results
  measurement.df <- rbind.fill(measurement.df, temp$measurement)
  sample.df <- rbind.fill(sample.df, temp$sample)
  
  ##CaCO3 + Texture + CAT exchange, column 43:46, 49
  if(verbose) cat('CaCO3, texture, cat exchange \n')
  #header.df[c(12, 43:46, 49),]
  sampleTemp <- data.df[, c(12, 43:46, 49)]
  names(sampleTemp) <- c('fieldID', header.df$measurement[c(43:46, 49)])
  temp <- processMethodBlock(methodNames=NULL,
                             sampleTemp=sampleTemp, 
                             unit.df=header.df[c(43:46, 49), c('measurement', 'unit')])
  #merge results
  measurement.df <- rbind.fill(measurement.df, temp$measurement)
  sample.df <- rbind.fill(sample.df, temp$sample)
  
  ##WPG2, columns 47 48
  if(verbose) cat('wpg2\n')
  #header.df[c(12, 47, 48),]
  sampleTemp <- data.df[, c(12, 47, 48)]
  names(sampleTemp) <- c('fieldID', header.df$measurement[c(47:48)])
  temp <- processMethodBlock(methodNames=header.df$measurement[47],
                             sampleTemp=sampleTemp, 
                             unit.df=header.df[c(47:48), c('measurement', 'unit')])
  #merge results
  measurement.df <- rbind.fill(measurement.df, temp$measurement)
  sample.df <- rbind.fill(sample.df, temp$sample)
  
  ##Al, columns 50, 52, 59, 60
  ##Fe, columns 53, 55, 59, 60
  ##Mn, columns 56, 58, 59 ??60
  #header.df[c(12, 50:60),]
  if(verbose) cat('Al Fe Mn\n')
  sampleTemp <- data.df[, c(12, 50:60)]
  names(sampleTemp) <- c('fieldID', header.df$measurement[c(50:60)])
  temp <- processMethodBlock(methodNames=header.df$measurement[60], 
                             unitName=header.df$measurement[59],
                             sampleTemp=sampleTemp)
  #merge results
  measurement.df <- rbind.fill(measurement.df, temp$measurement)
  sample.df <- rbind.fill(sample.df, temp$sample)
  
  #BC===============
  if(verbose) cat('Bc\n')
  ##BC, columns 61:67
  #header.df[c(12, 61:67),]
  sampleTemp <- data.df[, c(12, 61:67)]
  names(sampleTemp) <- c('fieldID', header.df$measurement[c(61:67)])
  temp <- processMethodBlock(methodNames=header.df$measurement[67], 
                             unitName=header.df$measurement[66],
                             sampleTemp=sampleTemp)
  #merge results
  measurement.df <- rbind.fill(measurement.df, temp$measurement)
  sample.df <- rbind.fill(sample.df, temp$sample)
  
  #CEC===========
  if(verbose) cat('CEC\n')
  ##CEC_h, columns 68:71
  #header.df[c(12, 68:71),]
  sampleTemp <- data.df[, c(12, 68:71)]
  names(sampleTemp) <- c('fieldID', header.df$measurement[c(68:71)])
  temp <- processMethodBlock(methodNames=NULL, 
                             unitName=header.df$measurement[71],
                             sampleTemp=sampleTemp)
  #merge results
  measurement.df <- rbind.fill(measurement.df, temp$measurement)
  sample.df <- rbind.fill(sample.df, temp$sample)
  #BS=======
  if(verbose) cat('BS\n')
  ##bs, columns 72:73
  sampleTemp <- data.df[, c(12, 72:73)]
  names(sampleTemp) <- c('fieldID', header.df$measurement[c(72:73)])
  temp <- processMethodBlock(methodNames=NULL, 
                             unitName=NULL,
                             unit.df=header.df[c(72:73), c('measurement', 'unit')],
                             sampleTemp=sampleTemp)
  #merge results
  measurement.df <- rbind.fill(measurement.df, temp$measurement)
  sample.df <- rbind.fill(sample.df, temp$sample)
  
  #Metal==========
  if(verbose) cat('metals \n')
  ##metal_ext, columns 74:77
  sampleTemp <- data.df[, c(12, 74:77)]
  names(sampleTemp) <- c('fieldID', header.df$measurement[c(74:77)])
  temp <- processMethodBlock(methodNames=header.df$measurement[77], 
                             unitName=header.df$measurement[76],
                             unit.df=NULL,
                             sampleTemp=sampleTemp)
  #merge results
  measurement.df <- rbind.fill(measurement.df, temp$measurement)
  sample.df <- rbind.fill(sample.df, temp$sample)
  
  #P ============
  if(verbose) cat('P\n')
  ##P, columns 78:83
  sampleTemp <- data.df[, c(12, 78:83)]
  names(sampleTemp) <- c('fieldID', header.df$measurement[c(78:83)])
  temp <- processMethodBlock(methodNames=header.df$measurement[83], 
                             unitName=header.df$measurement[82],
                             unit.df=NULL,
                             sampleTemp=sampleTemp)
  #merge results
  measurement.df <- rbind.fill(measurement.df, temp$measurement)
  sample.df <- rbind.fill(sample.df, temp$sample)
  
  #Roots + Isotope + texture + locator
  ##Root, column 84:85 84is TEXT move to field description
  ##Isotope, 86:93
  ##textureClass, 94 TEXT move to field description
  ##locator, 95 TEXT move to field description
  if(verbose) cat('roots isotope texture locator\n')
  sampleTemp <- data.df[, c(12, 85:93)]
  names(sampleTemp) <- c('fieldID', header.df$measurement[c(85:93)])
  temp <- processMethodBlock(methodNames=NULL, 
                             unitName=NULL,
                             unit.df=header.df[c(85:93),],
                             sampleTemp=sampleTemp)
  #merge results
  measurement.df <- rbind.fill(measurement.df, temp$measurement)
  sample.df <- rbind.fill(sample.df, temp$sample)
 
  if(verbose) cat('done with processWorksheet\n')
  return(list(study=study.df, field=field.df, lab=lab.df, 
              fieldTreatment=fieldTreatment.df, labTreatment=labTreatment.df, 
              measurement=measurement.df, samples=sample.df)) 
}